test/e2e/network/endpointslice.go

```diff
@@ -299,6 +303,180 @@ var _ = SIGDescribe("EndpointSlice", func() {
+									ContainerPort: int32(3000),
+								}},
+							},
+						},
+					},
+				},
+			},
+		}, metav1.CreateOptions{})
+		framework.ExpectNoError(err, "Error creating %s/%s Deployment", f.Namespace.Name, deployment.Name)
+
+		ownerRef := metav1.NewControllerRef(deployment, schema.GroupVersionKind{Group: "apps", Version: "v1", Kind: "Deployment"})
+		svc := createServiceReportErr(cs, f.Namespace.Name, &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            "example-int-port",
+				Labels:          sharedLabels,
+				OwnerReferences: []metav1.OwnerReference{*ownerRef},
+			},
+			Spec: v1.ServiceSpec{
+				// No need to wait for pods to be ready for this test
+				PublishNotReadyAddresses: true,
+				Selector:                 sharedLabels,
+				Ports: []v1.ServicePort{{
+					Name:       "example",
+					Port:       80,
+					TargetPort: intstr.FromInt(3000),
+					Protocol:   v1.ProtocolTCP,
+				}},
+			},
+		})
+		lSelector := discoveryv1beta1.LabelServiceName + "=" + svc.Name
+		maxResourceVersion := waitForNumEndpointSlices(cs, f.Namespace.Name, lSelector, 1)
+
+		deletePropagationForeground := metav1.DeletePropagationForeground
+		err = cs.AppsV1().Deployments(f.Namespace.Name).Delete(context.TODO(), deployment.Name, metav1.DeleteOptions{
+			PropagationPolicy: &deletePropagationForeground,
+		})
+		framework.ExpectNoError(err, "Error deleting %s/%s Deployment", f.Namespace.Name, deployment.Name)
+
+		// There's some delay between when a Service is deleted and the
+		// EndpointSlice informer is updated with that information. This
+		// allows for up to 2 EndpointSlice creations within 15 seconds of
+		// deletion.
+		expectMinimalEndpointSliceCreation(cs, f.Namespace.Name, lSelector, maxResourceVersion, 2, 5*time.Second, 15*time.Second)
+
+		// All EndpointSlices should be garbage collected within 5 minutes.
+		expectEndpointSliceGarbageCollection(cs, f.Namespace.Name, lSelector, 5*time.Minute)
+	})
+
+	// Ensure that garbage collection works well when a Service is deleted and
+	// its EndpointSlices are being rapidly updated.
+	// For more context: https://github.com/kubernetes/kubernetes/issues/91287
+	ginkgo.It("should not get in the way of Service garbage collection during rapid churn [Serial]", func() {
+		sharedLabels := map[string]string{"test-garbage-collection": "for-service"}
+
+		svc := createServiceReportErr(cs, f.Namespace.Name, &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "example-int-port",
+			},
+			Spec: v1.ServiceSpec{
+				// No need to wait for pods to be ready for this test
+				PublishNotReadyAddresses: true,
+				Selector:                 sharedLabels,
+				Ports: []v1.ServicePort{{
+					Name:       "example",
+					Port:       80,
+					TargetPort: intstr.FromInt(3000),
+					Protocol:   v1.ProtocolTCP,
+				}},
+			},
+		})
+
+		// Create an initial set of 100 test pods for initial EndpointSlice.
+		createTestPods(podClient, 0, 100, sharedLabels)
```

Comment:
are we sure we'll always have 100 pods' worth of capacity here? @smarterclayton, I have vague memories of you writing an e2e that tried to scale down to cluster capacity at small sizes... do you remember that?

Reply:
Certainly we have enforced in tests that we have to be able to test on just a few small nodes (like 2, with maybe 1-2 cores each).  I *think* we already have tests at 100 pods in serial, but maybe not in parallel.

Reply:
Looks like there's at least 1 serial e2e test running 100 pods _per node_. (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/node/kubelet_perf.go#L254 and https://github.com/kubernetes/kubernetes/blob/master/test/e2e/node/kubelet_perf.go#L69).

Reply:
that also has a Feature tag (`Feature:RegularResourceUsageTracking`) so I don't think it runs in default test configurations

Reply:
Good catch, I can remove this test if it's an issue. Or maybe I could add a `[Feature:EndpointSlice]` tag here to have a similar limitation. It would be nice if we could have some kind of e2e test coverage of an EndpointSlice being filled up and overflowing into another one, but maybe we just need to rely on unit tests for that? It seems like the only alternative would be a separate set of e2e tests that ran the controller with less endpoints per slice.

Reply:
if we're wanting to backport this fix (and I think we should), I would probably drop this e2e. exercising the controller in an integration test with a lower endpoints-per-slice setting with a simulated set of ready pods controlled by the test might be more sensible... you could exercise the apiserver/gc/endpoints interactions without actually starting 100 processes on a remote node somewhere

Reply:
Good call, an integration test does seem ideal here. As far as the e2e test, I tried one more approach that I think might still be helpful here. Instead of using lots of pods to get multiple EndpointSlices, I used named ports with unique port numbers on each pod. That means that each Pod will get its own EndpointSlice. The test now uses a total of 6 Pods and ends up with as many as 6 EndpointSlices (depending on the timing of the Service deletion). I think it's still potentially helpful to have test coverage ensuring that Service deletion can still work when the EndpointSlice controller would otherwise actively be trying to create EndpointSlices, and maybe more importantly test coverage ensuring that multiple EndpointSlices can be reliably garbage collected. With all that said, I'm still open to removing this entirely if it would help the change be cherry picked back.

Reply:
My recommendation generally is to test at the lowest level where a flawed interaction can be reproduced. At first glance, that would seem to be in an integration test (though I'm not sure how hard mocking the "ready" pods driving the endpointslice controller would be). If that proves difficult, the slimmed down e2e that is 1:1 pod:endpoint slice sounds like a good backup plan.